# import

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

In [2]:
file_path = 'C:/Users/Demi/Desktop/TB2/Mini project/df.csv'
data = pd.read_csv(file_path)
data = data.drop('Unnamed: 0', axis=1)
data

Account No       Balance   Amount  Third Party Account No  \
0       678330503.0   2971.000000  1584.00                     NaN   
1       472213568.0   3792.000000  1950.00                     NaN   
2       472213568.0   3012.000000  -780.00             283027736.0   
3       283027736.0   1787.000000   780.00             472213568.0   
4       624500124.0   3226.000000  1825.00                     NaN   
...             ...           ...      ...                     ...   
230190  581655972.0  45935.206861   -41.06                     NaN   
230191  786141370.0   -244.837500   -62.35                     NaN   
230192  824916823.0   9709.172159   -32.94                     NaN   
230193  366550080.0  26834.165794   -19.25                     NaN   
230194  259015404.0   1145.349837   -37.64                     NaN   

          Third Party Name             DateTime  
0       Westport Care Home  2023-01-01 00:00:00  
1         Barbiee Boutique  2023-01-01 00:00:00  
2                      NaN  2023-01-01 00:00:00  
3                      NaN  2023-01-01 00:00:00  
4                 Fat Face  2023-01-01 00:00:00  
...                    ...                  ...  
230190               Tesco  2023-12-06 20:54:00  
230191     Sainsbury Local  2023-12-06 20:55:00  
230192           Deliveroo  2023-12-06 21:05:00  
230193              Amazon  2023-12-06 21:13:00  
230194             JustEat  2023-12-06 21:32:00  

[230195 rows x 6 columns]

In [3]:
income = data[data['Amount'] > 0]
income = income.sort_values(by=['Account No', 'DateTime'])
income

Account No      Balance  Amount  Third Party Account No  \
20203   101531259.0   427.437950  1542.5                     NaN   
38427   101531259.0   713.096750  1542.5                     NaN   
56651   101531259.0  1109.055050  1542.5                     NaN   
74362   101531259.0  1299.455700  1542.5                     NaN   
93250   101531259.0  1479.357600  1542.5                     NaN   
...             ...          ...     ...                     ...   
133408  999752672.0  2499.562844  2708.0                     NaN   
154652  999752672.0  2691.089544  2708.0                     NaN   
176132  999752672.0  2769.420944  2708.0                     NaN   
198500  999752672.0  3337.786894  2708.0                     NaN   
220368  999752672.0  3519.263894  2708.0                     NaN   

       Third Party Name             DateTime  
20203        Hobbycraft  2023-01-31 00:00:00  
38427        Hobbycraft  2023-02-28 00:00:00  
56651        Hobbycraft  2023-03-31 00:00:00  
74362        Hobbycraft  2023-04-30 00:00:00  
93250        Hobbycraft  2023-05-31 00:00:00  
...                 ...                  ...  
133408              CPA  2023-07-31 00:00:00  
154652              CPA  2023-08-31 00:00:00  
176132              CPA  2023-09-30 00:00:00  
198500              CPA  2023-10-31 00:00:00  
220368              CPA  2023-11-30 00:00:00  

[10416 rows x 6 columns]

In [4]:
income['DateTime'] = pd.to_datetime(income['DateTime'])

# Split the datetime column into two separate columns for date and time
income['Date'] = income['DateTime'].dt.date
income['Time'] = income['DateTime'].dt.time
income

Account No      Balance  Amount  Third Party Account No  \
20203   101531259.0   427.437950  1542.5                     NaN   
38427   101531259.0   713.096750  1542.5                     NaN   
56651   101531259.0  1109.055050  1542.5                     NaN   
74362   101531259.0  1299.455700  1542.5                     NaN   
93250   101531259.0  1479.357600  1542.5                     NaN   
...             ...          ...     ...                     ...   
133408  999752672.0  2499.562844  2708.0                     NaN   
154652  999752672.0  2691.089544  2708.0                     NaN   
176132  999752672.0  2769.420944  2708.0                     NaN   
198500  999752672.0  3337.786894  2708.0                     NaN   
220368  999752672.0  3519.263894  2708.0                     NaN   

       Third Party Name   DateTime        Date      Time  
20203        Hobbycraft 2023-01-31  2023-01-31  00:00:00  
38427        Hobbycraft 2023-02-28  2023-02-28  00:00:00  
56651        Hobbycraft 2023-03-31  2023-03-31  00:00:00  
74362        Hobbycraft 2023-04-30  2023-04-30  00:00:00  
93250        Hobbycraft 2023-05-31  2023-05-31  00:00:00  
...                 ...        ...         ...       ...  
133408              CPA 2023-07-31  2023-07-31  00:00:00  
154652              CPA 2023-08-31  2023-08-31  00:00:00  
176132              CPA 2023-09-30  2023-09-30  00:00:00  
198500              CPA 2023-10-31  2023-10-31  00:00:00  
220368              CPA 2023-11-30  2023-11-30  00:00:00  

[10416 rows x 8 columns]

In [5]:
income['Time'].value_counts()

Time
00:00:00    7164
23:59:00    2343
17:08:00       6
17:04:00       6
17:11:00       6
            ... 
08:26:00       1
07:53:00       1
00:33:00       1
21:15:00       1
19:42:00       1
Name: count, Length: 592, dtype: int64

#### Most of transaction with Amonut>0 in fix time  
#### Separate to fix time and no fix time

In [6]:
condition1 = income['Time'].isin([pd.to_datetime('00:00').time(), pd.to_datetime('23:59').time()])
condition2 = ~income['Time'].isin([pd.to_datetime('00:00').time(), pd.to_datetime('23:59').time()])
income1 = income[condition1]
income1

Account No      Balance  Amount  Third Party Account No  \
20203   101531259.0   427.437950  1542.5                     NaN   
38427   101531259.0   713.096750  1542.5                     NaN   
56651   101531259.0  1109.055050  1542.5                     NaN   
74362   101531259.0  1299.455700  1542.5                     NaN   
93250   101531259.0  1479.357600  1542.5                     NaN   
...             ...          ...     ...                     ...   
133408  999752672.0  2499.562844  2708.0                     NaN   
154652  999752672.0  2691.089544  2708.0                     NaN   
176132  999752672.0  2769.420944  2708.0                     NaN   
198500  999752672.0  3337.786894  2708.0                     NaN   
220368  999752672.0  3519.263894  2708.0                     NaN   

       Third Party Name   DateTime        Date      Time  
20203        Hobbycraft 2023-01-31  2023-01-31  00:00:00  
38427        Hobbycraft 2023-02-28  2023-02-28  00:00:00  
56651        Hobbycraft 2023-03-31  2023-03-31  00:00:00  
74362        Hobbycraft 2023-04-30  2023-04-30  00:00:00  
93250        Hobbycraft 2023-05-31  2023-05-31  00:00:00  
...                 ...        ...         ...       ...  
133408              CPA 2023-07-31  2023-07-31  00:00:00  
154652              CPA 2023-08-31  2023-08-31  00:00:00  
176132              CPA 2023-09-30  2023-09-30  00:00:00  
198500              CPA 2023-10-31  2023-10-31  00:00:00  
220368              CPA 2023-11-30  2023-11-30  00:00:00  

[9507 rows x 8 columns]

In [7]:
income2 = income[condition2]
income2

Account No      Balance  Amount  Third Party Account No  \
33429   104832000.0    50.799341   290.0             587923845.0   
81173   104832000.0 -1212.478109   160.0             587923845.0   
19143   108931112.0   560.765200   950.0             646840694.0   
13926   113747882.0  1778.994750  1200.0             633345190.0   
7866    116169597.0  -540.449900   200.0             846550094.0   
...             ...          ...     ...                     ...   
101766  993238248.0  -654.680487    30.0             448279144.0   
140912  993238248.0  -559.580487   230.0             448279144.0   
18407   993663752.0   221.218492    90.0             672994751.0   
89899   998405607.0   101.278000   590.0             388862238.0   
122904  998405607.0 -1165.844000  1360.0             388862238.0   

       Third Party Name            DateTime        Date      Time  
33429               NaN 2023-02-15 06:57:00  2023-02-15  06:57:00  
81173               NaN 2023-05-02 12:18:00  2023-05-02  12:18:00  
19143               NaN 2023-01-28 13:38:00  2023-01-28  13:38:00  
13926               NaN 2023-01-17 04:20:00  2023-01-17  04:20:00  
7866                NaN 2023-01-05 17:11:00  2023-01-05  17:11:00  
...                 ...                 ...         ...       ...  
101766              NaN 2023-06-06 04:13:00  2023-06-06  04:13:00  
140912              NaN 2023-08-03 19:42:00  2023-08-03  19:42:00  
18407               NaN 2023-01-27 06:07:00  2023-01-27  06:07:00  
89899               NaN 2023-05-22 17:56:00  2023-05-22  17:56:00  
122904              NaN 2023-07-08 22:31:00  2023-07-08  22:31:00  

[909 rows x 8 columns]

# Salary1

In [8]:
average_values_df = income1.groupby(['Account No', 'Amount', ])['Amount'].count().reset_index(name='Count')
average_values_df.columns = ['Account No', 'Amount', 'Count']
average_values_df

Account No  Amount  Count
0    101531259.0  1542.5     11
1    104832000.0  1675.0     12
2    105375973.0   900.0     11
3    106601471.0  1675.0     11
4    108481285.0   640.0     12
..           ...     ...    ...
824  995615876.0  1733.0     11
825  996042490.0  3750.0     11
826  998390769.0  1625.0     12
827  998405607.0   450.0     11
828  999752672.0  2708.0     11

[829 rows x 3 columns]

In [36]:
average_values[average_values > 12]

Series([], Name: Amount, dtype: int64)

In [37]:
average_values[average_values > 9]

Account No   Amount
101531259.0  1542.5    11
104832000.0  1675.0    12
105375973.0  900.0     11
106601471.0  1675.0    11
108481285.0  640.0     12
                       ..
995615876.0  1733.0    11
996042490.0  3750.0    11
998390769.0  1625.0    12
998405607.0  450.0     11
999752672.0  2708.0    11
Name: Amount, Length: 829, dtype: int64

In [38]:
average_values[average_values < 10]

Series([], Name: Amount, dtype: int64)

#### All fix time transaction in same account and same amount happen 10~12 times

In [30]:
data['Account No'].nunique()

976

In [11]:
income1['Account No'].nunique()

810

#### There is 976 Accounts in origin dataset, and 810 Accounts have monthly income  
#### In 810 accounts, 791 accounts have one type monthly income and 19 Accounts have two types monthly inocme

In [12]:
duplicates = average_values_df[average_values_df.duplicated('Account No', keep=False)].sort_values('Account No')
duplicates

Account No  Amount  Count
4    108481285.0   640.0     12
5    108481285.0  1584.0     11
47   145387690.0   590.0     11
48   145387690.0  1675.0     12
50   146909202.0  1270.0     12
51   146909202.0  1287.0     12
147  259015404.0   600.0     12
148  259015404.0  1633.0     12
154  264550521.0  1675.0     12
153  264550521.0  1010.0     11
176  287480331.0   940.0     12
177  287480331.0  1596.0     11
197  310375875.0   580.0     12
198  310375875.0  1675.0     11
242  364357875.0   840.0     11
243  364357875.0  1842.5     12
279  407574996.0  1609.0     12
278  407574996.0  1300.0     12
327  456246119.0  2083.3     12
326  456246119.0   610.0     12
357  492215823.0  1360.0     11
358  492215823.0  1675.0     11
437  572384506.0  1390.0     12
438  572384506.0  1683.0     11
452  587225252.0   800.0     11
453  587225252.0  2000.0     11
471  609939116.0   860.0     11
472  609939116.0  1706.0     12
516  653667523.0  1675.0     10
515  653667523.0   560.0     12
545  678330503.0   660.0     11
546  678330503.0  1584.0     12
627  766941757.0  1230.0     12
628  766941757.0  1675.0     11
747  899742927.0  1000.0     12
748  899742927.0  1842.5     12
794  960208650.0  1160.0     11
795  960208650.0  1950.0     12

In [13]:
income1['Date'].value_counts()

Date
2023-01-31    422
2023-04-30    422
2023-06-30    422
2023-07-31    422
2023-08-31    422
2023-09-30    422
2023-11-30    422
2023-02-28    421
2023-03-31    421
2023-05-31    421
2023-10-31    421
2023-01-01    396
2023-12-01    295
2023-11-01    295
2023-10-01    295
2023-09-01    295
2023-07-01    295
2023-05-01    295
2023-03-01    295
2023-06-01    294
2023-04-01    294
2023-02-01    294
2023-08-01    294
2023-06-29    112
2023-10-30    112
2023-09-29    112
2023-08-30    112
2023-07-30    112
2023-03-30    112
2023-05-30    112
2023-04-29    112
2023-02-27    112
2023-01-30    112
2023-11-29    112
Name: count, dtype: int64

#### In income1 dataset and the Timestamp is 00:00 and 23:59, these are definitely the salary  
#### It happens only at start or end of every month, every same account and same amount have 10~12 records, which maybe due to missing record or December.

# Salary2

In [25]:
average_values_df2 = income2.groupby(['Account No', 'Amount', ])['Amount'].count()
average_values_df2

Account No   Amount
104832000.0  160.0     1
             290.0     1
108931112.0  950.0     1
113747882.0  1200.0    1
116169597.0  200.0     1
                      ..
993238248.0  30.0      1
             230.0     1
993663752.0  90.0      1
998405607.0  590.0     1
             1360.0    1
Name: Amount, Length: 894, dtype: int64

In [28]:
average_values_df2[average_values_df2 > 2]

Series([], Name: Amount, dtype: int64)

In [29]:
average_values_df2[average_values_df2 > 1]

Account No   Amount
122884111.0  1920.0    2
235663459.0  1220.0    2
248658670.0  190.0     2
             640.0     2
256045540.0  70.0      2
280098744.0  2060.0    2
293609137.0  240.0     2
310620484.0  970.0     2
329840194.0  10.0      2
496674181.0  420.0     2
596463612.0  1400.0    2
657528784.0  150.0     2
762896656.0  470.0     2
812012103.0  190.0     2
961077907.0  550.0     2
Name: Amount, dtype: int64

In [31]:
income2['Date'].value_counts()

Date
2023-05-07    24
2023-03-08    22
2023-01-03    20
2023-06-13    18
2023-03-30    17
              ..
2023-10-15     1
2023-06-01     1
2023-09-14     1
2023-03-26     1
2023-07-08     1
Name: count, Length: 174, dtype: int64

#### In not fix time income, most of transaction happen only 1~2 times, so I think we could 

# Salary1 - company and personal store

In [8]:
salary1 = income1.copy()
salary1.drop(['Balance', 'DateTime', 'Date', 'Time'], axis=1, inplace=True)
salary1

Account No  Amount  Third Party Account No Third Party Name
20203   101531259.0  1542.5                     NaN       Hobbycraft
38427   101531259.0  1542.5                     NaN       Hobbycraft
56651   101531259.0  1542.5                     NaN       Hobbycraft
74362   101531259.0  1542.5                     NaN       Hobbycraft
93250   101531259.0  1542.5                     NaN       Hobbycraft
...             ...     ...                     ...              ...
133408  999752672.0  2708.0                     NaN              CPA
154652  999752672.0  2708.0                     NaN              CPA
176132  999752672.0  2708.0                     NaN              CPA
198500  999752672.0  2708.0                     NaN              CPA
220368  999752672.0  2708.0                     NaN              CPA

[9507 rows x 4 columns]

In [9]:
salary1.drop_duplicates(inplace=True)

In [10]:
salary1

Account No  Amount  Third Party Account No Third Party Name
20203  101531259.0  1542.5                     NaN       Hobbycraft
93     104832000.0  1675.0                     NaN     A Yarn Story
20587  105375973.0   900.0             709632525.0              NaN
20298  106601471.0  1675.0                     NaN       Hobbycraft
78     108481285.0   640.0             913801979.0              NaN
...            ...     ...                     ...              ...
20217  995615876.0  1733.0                     NaN          Matalan
20221  996042490.0  3750.0                     NaN  Lloyds Pharmacy
336    998390769.0  1625.0                     NaN       Kings Arms
20580  998405607.0   450.0             388862238.0              NaN
20432  999752672.0  2708.0                     NaN              CPA

[829 rows x 4 columns]

In [11]:
salary_company = salary1.dropna(subset=['Third Party Name'], inplace=False)
salary_company

Account No  Amount  Third Party Account No Third Party Name
20203  101531259.0  1542.5                     NaN       Hobbycraft
93     104832000.0  1675.0                     NaN     A Yarn Story
20298  106601471.0  1675.0                     NaN       Hobbycraft
20334  108481285.0  1584.0                     NaN  Happy Days Home
20392  108563213.0  3749.0                     NaN              LBG
...            ...     ...                     ...              ...
20521  994947671.0  1788.0                     NaN  Grand Union BJJ
20217  995615876.0  1733.0                     NaN          Matalan
20221  996042490.0  3750.0                     NaN  Lloyds Pharmacy
336    998390769.0  1625.0                     NaN       Kings Arms
20432  999752672.0  2708.0                     NaN              CPA

[616 rows x 4 columns]

In [12]:
salary_personal_store = salary1.dropna(subset=['Third Party Account No'], inplace=False)
salary_personal_store

Account No  Amount  Third Party Account No Third Party Name
20587  105375973.0   900.0             709632525.0              NaN
78     108481285.0   640.0             913801979.0              NaN
20564  108812033.0   830.0             386277162.0              NaN
182    117308635.0  1200.0             255814263.0              NaN
20276  119943495.0   400.0             352899149.0              NaN
...            ...     ...                     ...              ...
20365  985188462.0   830.0             593378953.0              NaN
20619  987656636.0  2670.0             450896297.0              NaN
235    993238248.0   540.0             448279144.0              NaN
20591  993663752.0   670.0             672994751.0              NaN
20580  998405607.0   450.0             388862238.0              NaN

[213 rows x 4 columns]

#### There is 829 fix time income, and 616 income from company, 213 from personal store

In [14]:
salary1_sorted = salary1.sort_values(by='Account No')

In [15]:
accounts_with_two_income = salary1_sorted.groupby('Account No').filter(lambda x: len(x) == 2)
accounts_with_two_income

Account No  Amount  Third Party Account No    Third Party Name
78     108481285.0   640.0             913801979.0                 NaN
20334  108481285.0  1584.0                     NaN     Happy Days Home
56     145387690.0  1675.0                     NaN           The Crown
20535  145387690.0   590.0             221868392.0                 NaN
371    146909202.0  1270.0             741269136.0                 NaN
368    146909202.0  1287.0                     NaN       Fitted Stitch
95     259015404.0  1633.0                     NaN          North Face
96     259015404.0   600.0             146383025.0                 NaN
20269  264550521.0  1010.0             151278285.0                 NaN
351    264550521.0  1675.0                     NaN               Tesco
122    287480331.0   940.0             585596026.0                 NaN
20224  287480331.0  1596.0                     NaN     Happy Days Home
287    310375875.0   580.0             903663593.0                 NaN
20491  310375875.0  1675.0                     NaN           Sainsbury
345    364357875.0  1842.5                     NaN            Cass Art
20445  364357875.0   840.0             833499132.0                 NaN
16     407574996.0  1609.0                     NaN            Fat Face
19     407574996.0  1300.0             172084472.0                 NaN
183    456246119.0  2083.3                     NaN          Kings Arms
186    456246119.0   610.0             489964679.0                 NaN
20543  492215823.0  1360.0             492058070.0                 NaN
20293  492215823.0  1675.0                     NaN    Stitch By Stitch
273    572384506.0  1390.0             647936800.0                 NaN
20248  572384506.0  1683.0                     NaN      Collector Cave
20442  587225252.0  2000.0                     NaN             Matalan
20710  587225252.0   800.0             425964226.0                 NaN
419    609939116.0  1706.0                     NaN  Westport Care Home
20216  609939116.0   860.0             627917561.0                 NaN
479    653667523.0   560.0             939972497.0                 NaN
20245  653667523.0  1675.0                     NaN      Collector Cave
0      678330503.0  1584.0                     NaN  Westport Care Home
20425  678330503.0   660.0             355733816.0                 NaN
335    766941757.0  1230.0             355940285.0                 NaN
20232  766941757.0  1675.0                     NaN    Stitch By Stitch
279    899742927.0  1000.0             479366718.0                 NaN
280    899742927.0  1842.5                     NaN         Pets Corner
436    960208650.0  1950.0                     NaN    Barbiee Boutique
20538  960208650.0  1160.0             535804582.0                 NaN

#### Account have two incomes, one from company, the other from personal transaction